### PR TITLE
implemented the spring security to the project

### PR DESCRIPTION
**Pull request by ivan0dyatlyukkk**

### CODE REVIEW

File: src/main/java/org/diatliuk/bookstore/controller/BookController.java [hunk 1]
```diff
@@ -29,24 +30,28 @@ public class BookController {
     private final BookService bookService;
 
     @GetMapping
+    @PreAuthorize("hasRole('ROLE_USER')")
```
Comment (DmytroV95) [inline src/main/java/org/diatliuk/bookstore/controller/BookController.java]: I think you should open access for admin too. Think about others endpoints.

File: src/main/java/org/diatliuk/bookstore/dto/user/UserLoginRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import io.swagger.v3.oas.annotations.media.Schema;
+import jakarta.validation.constraints.NotBlank;
+import lombok.Data;
+import org.hibernate.validator.constraints.Length;
+
+@Data
+public class UserLoginRequestDto {
+    @NotBlank
+    @Schema(example = "new@example.com")
+    private String email;
```
Comment (DmytroV95) [inline src/main/java/org/diatliuk/bookstore/dto/user/UserLoginRequestDto.java:14]: To my mind, using a custom annotation with regex pattern covering more cases for different emails patterns. Also think about if the field is empty. Maybe add more cases to check data validation.

File: src/main/java/org/diatliuk/bookstore/dto/user/UserRegistrationRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+@Data
+@FieldMatch.List({
+        @FieldMatch(
+                field = "password",
+                fieldMatch = "repeatPassword",
+                message = "The password and repeat password aren't the same"
+        )
+})
+public class UserRegistrationRequestDto {
```
Comment (DmytroV95) [inline src/main/java/org/diatliuk/bookstore/dto/user/UserRegistrationRequestDto.java:18]: What if this fields is null?

File: src/main/java/org/diatliuk/bookstore/model/Role.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import lombok.Data;
+import org.diatliuk.bookstore.enums.RoleName;
+
+@Data
+@Entity
+@Table(name = "roles")
```
Comment (DmytroV95) [inline src/main/java/org/diatliuk/bookstore/model/Role.java:16]:
```suggestion
@Table(name = "role")
```

File: src/main/java/org/diatliuk/bookstore/model/Role.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+public class Role {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Enumerated(EnumType.STRING)
+    private RoleName name;
```
Comment (DmytroV95) [inline src/main/java/org/diatliuk/bookstore/model/Role.java:24]: This field should be not null and unique.

File: src/main/java/org/diatliuk/bookstore/model/User.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+import org.hibernate.annotations.Where;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
+
+@Data
+@Entity
+@SQLDelete(sql = "UPDATE users SET is_deleted = true WHERE id = ?")
+@Where(clause = "is_deleted = false")
+@Table(name = "users")
```
Comment (DmytroV95) [inline src/main/java/org/diatliuk/bookstore/model/User.java:29]:
```suggestion
@Table(name = "user")
```

File: src/main/java/org/diatliuk/bookstore/security/JwtUtil.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.Date;
+import java.util.function.Function;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+@Component
+public class JwtUtil {
+    @Value("${jwt.expiration}")
+    private Long expiration;
+    private Key secret;
```
Comment (DmytroV95) [inline src/main/java/org/diatliuk/bookstore/security/JwtUtil.java]:
```suggestion
    private final Key secret;
```

File: src/main/java/org/diatliuk/bookstore/service/impl/UserServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    private static final RoleName DEFAULT_ROLE = RoleName.ROLE_USER;
+    private final UserRepository userRepository;
+    private final RoleRepository roleRepository;
+    private final PasswordEncoder passwordEncoder;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserLoginResponseDto login(UserLoginRequestDto requestDto) {
+        return null;
+    }
```
Comment (DmytroV95) [inline src/main/java/org/diatliuk/bookstore/service/impl/UserServiceImpl.java]: Do you need it?

File: src/main/resources/db/changelog/changes/04-create-roles-table.yaml [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+              - column:
+                  name: name
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+                    unique: true
```
Comment (DmytroV95) [inline src/main/resources/db/changelog/changes/04-create-roles-table.yaml]: No empty line.

File: src/main/resources/db/changelog/changes/05-create-users_roles-table.yaml [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+      changes:
+        - createTable:
+            tableName: users_roles
+            columns:
+              - column:
+                  name: user_id
+                  type: bigint
+              - column:
+                  name: role_id
+                  type: bigint
```
Comment (DmytroV95) [inline src/main/resources/db/changelog/changes/05-create-users_roles-table.yaml]: No empty line.

**Review by DmytroV95 — CHANGES_REQUESTED** [review]
GJ! Let's improve your solution!

File: src/main/java/org/diatliuk/bookstore/dto/user/UserLoginRequestDto.java (hunk 2 above)
Comment (ivan0dyatlyukkk) [reply src/main/java/org/diatliuk/bookstore/dto/user/UserLoginRequestDto.java:14]: Yes, you're right. My bad, I forgot to specify

File: src/main/java/org/diatliuk/bookstore/model/Role.java (hunk 4 above)
Comment (ivan0dyatlyukkk) [reply src/main/java/org/diatliuk/bookstore/model/Role.java:16]: I guess that it's not a problem, and it depends on naming conventions in a team

**Review by DmytroV95 — APPROVED** [review]
LGTM!

File: src/main/java/org/diatliuk/bookstore/security/CustomUserDetailsService.java [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+
+@Component
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        return userRepository.findByEmail(username)
+                .orElseThrow(() -> new EntityNotFoundException("Can't find a user by email: "
```
Comment (andrii-hoienko) [inline src/main/java/org/diatliuk/bookstore/security/CustomUserDetailsService.java]: Cant find by email or by user name?

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Very good, one minor comment left

**Review by andrii-hoienko — APPROVED** [review]

**Pull request merged**